Lib/glob.py

```diff
@@ -310,3 +157,229 @@ def translate(pat, *, recursive=False, include_hidden=False, seps=None):
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                if fd is not None:
+                    prefix = _add_trailing_slash(path)
+                with os.scandir(arg) as scandir_it:
+                    entries = list(scandir_it)
+                for entry in entries:
+                    if match is None or match(entry.name):
+                        try:
+                            if entry.is_dir():
+                                entry_path = entry.path
+                                if fd is not None:
+                                    entry_path = prefix + entry_path
+                                yield from select_next(entry_path, entry.name, fd, True)
+                        except OSError:
+                            pass
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    else:
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                prefix = _add_trailing_slash(path)
+                for name in os.listdir(arg):
+                    if match is None or match(name):
+                        yield prefix + name
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    return select_wildcard
+
+
+def _recursive_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects a given path and all its children,
+    recursively, filtering by pattern.
+    """
+    while parts and parts[0] == '**':
+        parts = parts[1:]
+    while parts and parts[0] not in _special_parts:
+        # Consume next non-special component (used to build regex).
+        part += os.path.sep + parts[0]
+        parts = parts[1:]
+
+    if include_hidden and part == '**':
+        match = None  # Skip generating a pattern that would match all inputs.
+    else:
+        match = _compile_pattern(part, recursive, include_hidden)
+
+    dir_only = bool(parts)
+    select_next = _selector(parts, recursive, include_hidden)
+
+    def select_recursive(path, rel_path, dir_fd, exists, match_pos=None):
+        if match_pos is None:
+            path = _add_trailing_slash(path)
+            rel_path = _add_trailing_slash(rel_path)
+            match_pos = len(path)
+
+        if match is None or match(path, match_pos):
+            yield from select_next(path, rel_path, dir_fd, exists)
+
+        close_fd = False
+        try:
+            arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+            if fd is not None:
+                prefix = _add_trailing_slash(path)
+            with os.scandir(arg) as scandir_it:
+                entries = list(scandir_it)
```

Comment:
Is there perhaps a goal to not have the number of open directory file descriptors match the recursion depth?

The downside of that is that it means file systems full of huge named trash files will cause excessive memory consumption during a glob when a wildcard is recursively searching a relevant bit of it.

Reply:
I believe the places you have code doing recursive `yield from`'s of selected subdirectory contents _could_ be done differently while limiting open directory file descriptors:

When looping over the iterator, yield non-dir matches immediately - and create a list of subdirectories that need checking, iterating over that and yielding from those only _after_ the current directory's iterator is exhausted. So that we're not expending temporary list space on non-directory file information.

That still has a potential for memory exhaustion, but smaller: only if enough total subdirectories exist in one place, not files.

From experience: Single directories with a huge number of long-name-length files are more naturally occurring than antagonistic directory trees full of huge named subdirectories. (not impossible, but there's only so much we can do if under a constraint to limit open fds - themselves _also_ an exhaustable resource)

Reply:
Sorry this isn't commented - I'll add some now. You're right that it's about avoiding exhausting file descriptors when globbing deep trees. I love the idea of reading just subdirectories into a list for later iteration - will try that!

Reply:
FWIW, the old code also reads every `os.scandir()` into a list.

Reply:
A follow-up PR makes sense to me.  Given the old code, I don't think past iglob users have ever enjoyed memory friendliness as a feature.